### ISSUE TITLE
Connection to server with invalid/no cert causes crash

### ISSUE THREAD

**Issue by patrikwlund** [versatica/OverSIP#83]
With standard settings and a letsencrypt cert, the server crashes when I'm unable to connect to some remote servers when using wss/WebRTC. These connections don't simpy fail and log something, they kill the process and say pretty much nothing about it. I then have to manually restart oversip.

I'm able to connect to my own server though, which has a valid cert.

I have no objections to the fact that it fails, it should. But surely there must be a somewhat graceful solution? Moreso than dying miserably atleast.

Could this simply be an error on my part, or is it perhaps the stud or eventmachine that's unstable?

Example address that has no cert:
sip:test.time@sip5060.net

Example address with invalid cert:
sip:dummy@sip.linphone.org

An excerpt from syslog, with all the logs I see during the life cycle:
http://pastebin.com/vS5LyHE9

Syslog in short:
Mar 16 05:05:39 debian oversip[3930]:  DEBUG: <Proxy proxy_out 446648> trying single target: tls:195.8.117.60:443
Mar 16 05:05:39 debian oversip[3930]:  DEBUG: <SIP TLS IPv4 client> TLS handshake not completed yet, waiting before sending the message
Mar 16 05:05:39 debian stud[10515]: {backend} Connection closed

**Comment by ibc** [comment]
Please clarify: do you mean that oversip crashes when it starts a TLS connection to a remote SIP+TLS server? is that?

I assume the answer is yes, and sadly I must say that TLS support in EventMachine is really poor and fragile. oversip comes with a built-in *stud* TLS server that is used to translate TLS to TCP for **incoming** TLS connections. But when it comes to initiate a TLS connection with a remote server, oversip uses the EventMachine ugly TLS stack, and yes.

Said that, OverSIP uses a fork of EventMachine called EventMachine-LE, but no special improvements have been done in the TLS layer.

At the same time, the original [EventMachine](https://github.com/eventmachine/eventmachine) project is being actively maintained so I suggest you the following:

* Install the official EventMachine (master/latest version) and use its Ruby API to initiate a TLS connection to those SIP+TLS servers above. It would probably crash so you can properly report it into the EventMachine project.

I may help on this, but currently I don't have too much time.

**Comment by patrikwlund** [comment]
That's correct. Oversip crashes.

I'll look into EventMachine and see if I can reproduce the problem there (which seems probable).

Thanks for the quick reply and explanation.

**Comment by patrikwlund** [comment]
Well that was easy, but somewhat confusing.

When testing EventMachine-LE I was unable to open _any_ connections with TLS, and instead got this error:
`"Encryption not available on this event-machine"`

So I uninstalled and then reinstalled
libssl-dev

And it simply works now. It fails gracefully, as expected. Not sure what went wrong here, but it was probably my mistake somehow. I'm 100% certain I had it installed already though, and I was able to connect to my other server using TLS through Oversip before reinstalling.

I wouldn't call this solved, but since the solution was so simple it doesn't really matter imo.
Thanks.
